File: source/Caca.VM.Tests/ConformanceTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Xunit;
using Caca.VM.Handlers;

namespace Caca.VM.Tests
{
    /// <summary>
    /// Runs the C# reference VM against the shared, engine-agnostic corpus in
    /// /conformance/cases — see conformance/README.md. The Swift port
    /// (mac/CacaStudioMac/Tests/CacaVMKitTests) runs the exact same corpus via its
    /// own adapter; a case only counts once both agree on it.
    /// </summary>
    [Collection("VM")]
    public sealed class ConformanceTests
    {
        /// <summary>Console that replays a fixed list of stdin lines and captures all output.</summary>
        private sealed class ScriptedConsole : VConsole
        {
            private readonly System.Text.StringBuilder _output = new System.Text.StringBuilder();
            private readonly Queue<string> _lines;

            public ScriptedConsole(IEnumerable<string> lines) => _lines = new Queue<string>(lines);

            public string Output => _output.ToString();

            public override void Write(char ch) => _output.Append(ch);
            public override void Write(string text) => _output.Append(text);
            public override void WriteLine(string text)
            {
                _output.Append(text);
                _output.Append('\n');
            }

            public override byte Read() => _lines.Count > 0 && _lines.Peek().Length > 0
                ? (byte)_lines.Peek()[0]
                : (byte)0;

            public override string ReadLine() => _lines.Count > 0 ? _lines.Dequeue() : string.Empty;
        }

        /// <summary>
        /// Walks up from this test assembly's own directory to find the repo root
        /// (the one containing both "source" and "conformance"), rather than relying
        /// on the test runner's current working directory.
        /// </summary>
        private static string ConformanceCasesDirectory()
        {
            var dir = new DirectoryInfo(AppContext.BaseDirectory);
            while (dir != null && !Directory.Exists(Path.Combine(dir.FullName, "conformance", "cases")))
            {
                dir = dir.Parent;
            }

            if (dir == null)
            {
                throw new DirectoryNotFoundException(
                    "Could not locate conformance/cases by walking up from " + AppContext.BaseDirectory);
            }

            return Path.Combine(dir.FullName, "conformance", "cases");
        }

        public static IEnumerable<object[]> Cases()
        {
            string casesDir = ConformanceCasesDirectory();
            foreach (string dir in Directory.GetDirectories(casesDir).OrderBy(d => d))
            {
                yield return new object[] { Path.GetFileName(dir) };
            }
        }

        [Theory]
        [MemberData(nameof(Cases))]
        public void ConformanceCase_MatchesExpectedBehavior(string caseName)
        {
            string dir = Path.Combine(ConformanceCasesDirectory(), caseName);

            string source = File.ReadAllText(Path.Combine(dir, "source.cil"));

            string stdinPath = Path.Combine(dir, "stdin.txt");
            var input = File.Exists(stdinPath)
                ? File.ReadAllText(stdinPath).Split('\n')
                : Array.Empty<string>();

            string ramSizePath = Path.Combine(dir, "ramsize.txt");
            int ramSize = File.Exists(ramSizePath)
                ? int.Parse(File.ReadAllText(ramSizePath).Trim())
                : Globals.DefaultRamSize;

            string stdoutPath = Path.Combine(dir, "stdout.txt");
            string errorPath = Path.Combine(dir, "error.txt");
            bool hasStdout = File.Exists(stdoutPath);
            bool hasError = File.Exists(errorPath);

            Assert.True(hasStdout != hasError,
                $"{caseName}: exactly one of stdout.txt/error.txt must be present");

            var console = new ScriptedConsole(input);
            Caca.VM.Globals.console = console;
            Caca.VM.Globals.DebugMode = false;

            byte[] code = new global::Caca.VM.Compiler.Compiler(source).Compile();
            var vm = new VM(code, ramSize);

            if (hasStdout)
            {
                vm.Execute();
                string expected = File.ReadAllText(stdoutPath);
                Assert.Equal(expected, console.Output);
            }
            else
            {
                string expectedSubstring = File.ReadAllText(errorPath).Trim();
                var ex = Assert.Throws<Exception>(() => vm.Execute());
                Assert.Contains(expectedSubstring, ex.Message);
            }
        }
    }
}

File: cacaswift/Tests/CacaVMKitTests/ConformanceTests.swift
import XCTest
@testable import CacaVMKit

/// Runs CacaVMKit against the shared, engine-agnostic corpus in
/// CacaVM/conformance/cases — see docs/CIL-Conformance-Suite.md. The C# engine
/// (CacaVM/source/Caca.VM.Tests) runs the exact same corpus via its own
/// adapter; a case only counts once both agree on it.
final class ConformanceTests: XCTestCase {
    private static var conformanceCasesDirectory: URL {
        // #filePath is this source file's own path at compile time; walk up
        // from it to the repo root (cacaswift/Tests/CacaVMKitTests/ -> repo
        // root) rather than relying on the test bundle's working directory,
        // which SPM does not guarantee points anywhere useful.
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent() // CacaVMKitTests
            .deletingLastPathComponent() // Tests
            .deletingLastPathComponent() // cacaswift (repo root)
            .appendingPathComponent("CacaVM/conformance/cases")
    }

    private static func discoverCases() -> [String] {
        let fm = FileManager.default
        guard let names = try? fm.contentsOfDirectory(atPath: conformanceCasesDirectory.path) else {
            return []
        }
        return names.sorted()
    }

    func testConformanceCorpus() throws {
        let cases = Self.discoverCases()
        XCTAssertFalse(cases.isEmpty, "no conformance cases found at \(Self.conformanceCasesDirectory.path)")

        var failures: [String] = []

        for name in cases {
            let dir = Self.conformanceCasesDirectory.appendingPathComponent(name)
            do {
                try runCase(at: dir)
            } catch {
                failures.append("\(name): \(error)")
            }
        }

        XCTAssertTrue(failures.isEmpty, "Conformance failures:\n" + failures.joined(separator: "\n"))
    }

    private func runCase(at dir: URL) throws {
        let source = try String(contentsOf: dir.appendingPathComponent("source.cil"), encoding: .utf8)

        let stdinPath = dir.appendingPathComponent("stdin.txt")
        let input: [String] = (try? String(contentsOf: stdinPath, encoding: .utf8))
            .map { $0.split(separator: "\n", omittingEmptySubsequences: false).map(String.init) }
            ?? []

        let ramSizePath = dir.appendingPathComponent("ramsize.txt")
        let ramSize = (try? String(contentsOf: ramSizePath, encoding: .utf8))
            .flatMap { Int($0.trimmingCharacters(in: .whitespacesAndNewlines)) }
            ?? 1_048_576

        let expectedStdoutPath = dir.appendingPathComponent("stdout.txt")
        let expectedErrorPath = dir.appendingPathComponent("error.txt")
        let expectedStdout = try? String(contentsOf: expectedStdoutPath, encoding: .utf8)
        let expectedError = try? String(contentsOf: expectedErrorPath, encoding: .utf8)

        precondition(
            (expectedStdout != nil) != (expectedError != nil),
            "\(dir.lastPathComponent): exactly one of stdout.txt/error.txt must be present"
        )

        let console = BufferConsole(input: input)

        do {
            let code = try Assembler.assemble(source)
            let vm = VM(program: code, ramSize: ramSize, console: console)
            try vm.run()

            if let expectedError {
                XCTFail("\(dir.lastPathComponent): expected an error containing '\(expectedError.trimmingCharacters(in: .whitespacesAndNewlines))', but it ran to completion with output: \(console.output)")
                return
            }

            XCTAssertEqual(console.output, expectedStdout, dir.lastPathComponent)
        } catch {
            guard let expectedError else {
                XCTFail("\(dir.lastPathComponent): unexpected error \(error) (output so far: \(console.output))")
                return
            }

            let message = "\(error)"
            XCTAssertTrue(
                message.contains(expectedError.trimmingCharacters(in: .whitespacesAndNewlines)),
                "\(dir.lastPathComponent): error '\(message)' did not contain expected substring '\(expectedError)'"
            )
        }
    }
}
